File: Server~/src/tools/captureScreenshotTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Constants for the tool
const toolName = 'capture_screenshot';
const toolDescription = "Captures a screenshot from Unity's Scene or Game view";
const paramsSchema = z.object({
  viewType: z.enum(["scene", "game", "both"]).default("game").describe("Type of view to capture: 'scene', 'game', or 'both'"),
  width: z.number().default(0).describe("Width of the screenshot (0 for current size)"),
  height: z.number().default(0).describe("Height of the screenshot (0 for current size)"),
  saveToFile: z.boolean().default(false).describe("Whether to save the screenshot to a file"),
  filePath: z.string().optional().describe("Path where to save the screenshot (optional, auto-generated if not provided)")
});

/**
 * Creates and registers the Capture Screenshot tool with the MCP server
 * This tool allows capturing screenshots from Unity's Scene or Game view
 * 
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerCaptureScreenshotTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
  
  // Register this tool with the MCP server
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (args, extra): Promise<CallToolResult> => {
      try {
        const { 
          viewType = "game", 
          width = 0, 
          height = 0, 
          saveToFile = false,
          filePath 
        } = args;

        logger.info(`Capturing screenshot: viewType=${viewType}, width=${width}, height=${height}, saveToFile=${saveToFile}`);

        const result = await mcpUnity.sendRequest({
          method: "capture_screenshot",
          params: {
            viewType,
            width,
            height,
            saveToFile,
            filePath
          }
        });

        if (!result.success) {
          logger.error(`Screenshot capture failed: ${result.message}`);
          return {
            content: [
              {
                type: "text",
                text: `Failed to capture screenshot: ${result.message || "Unknown error"}`
              }
            ]
          };
        }

        // Handle the response based on what Unity returned
        if (result.type === "image" && result.data) {
          // Single screenshot - return as MCP image content
          const screenshot = result.data;
          return {
            content: [
              {
                type: "image",
                data: screenshot.data, // This should be the base64 data URL
                mimeType: "image/png"
              },
              {
                type: "text",
                text: `Captured ${screenshot.type} view screenshot (${screenshot.width}x${screenshot.height})`
              }
            ]
          };
        } else if (result.type === "images" && result.data?.screenshots) {
          // Multiple screenshots - return each as separate content
          const screenshots = result.data.screenshots;
          const content: any[] = [];
          
          for (const screenshot of screenshots) {
            content.push({
              type: "image",
              data: screenshot.data,
              mimeType: "image/png"
            });
          }
          
          content.push({
            type: "text",
            text: `Captured ${screenshots.length} screenshots: ${screenshots.map((s: any) => `${s.type} (${s.width}x${s.height})`).join(", ")}`
          });
          
          return { content };
        } else if (result.type === "files" && result.data?.files) {
          // Files saved - return paths
          const files = result.data.files;
          return {
            content: [
              {
                type: "text",
                text: `Screenshots saved to:\n${files.join("\n")}`
              }
            ]
          };
        }

        // Fallback response
        return {
          content: [
            {
              type: "text",
              text: result.message || "Screenshot captured successfully"
            }
          ]
        };
      } catch (error) {
        if (error instanceof McpUnityError) {
          logger.error(`Screenshot capture error: ${error.message}`);
          return {
            content: [
              {
                type: "text",
                text: `Error capturing screenshot: ${error.message}`
              }
            ]
          };
        }
        
        logger.error("Screenshot capture error:", error);
        return {
          content: [
            {
              type: "text",
              text: `Error capturing screenshot: ${error instanceof Error ? error.message : String(error)}`
            }
          ]
        };
      }
    }
  );
  
  logger.info(`Tool ${toolName} registered successfully`);
}

File: Editor/Tools/CaptureScreenshotTool.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// MCP tool for capturing screenshots from Unity
    /// </summary>
    public class CaptureScreenshotTool : McpToolBase
    {
        public CaptureScreenshotTool()
        {
            Name = "capture_screenshot";
            Description = "Captures a screenshot from Unity's Scene or Game view";
        }
        
        /// <summary>
        /// Execute the screenshot capture tool with the provided parameters synchronously
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        public override JObject Execute(JObject parameters)
        {
            try
            {
                // Parse parameters with reasonable defaults
                string viewType = parameters["viewType"]?.ToObject<string>() ?? "game";
                int width = parameters["width"]?.ToObject<int>() ?? 0;   // 0 = native game view resolution
                int height = parameters["height"]?.ToObject<int>() ?? 0; // 0 = native game view resolution
                bool saveToFile = parameters["saveToFile"]?.ToObject<bool>() ?? false;
                string filePath = parameters["filePath"]?.ToObject<string>();
                
                // Validate view type
                if (!IsValidViewType(viewType))
                {
                    return JObject.FromObject(new
                    {
                        content = new[]
                        {
                            new
                            {
                                type = "text",
                                text = $"Invalid viewType: {viewType}. Must be 'scene', 'game', or 'both'"
                            }
                        },
                        isError = true
                    });
                }
                
                Dictionary<string, object> result;
                
                if (saveToFile)
                {
                    result = ScreenshotCapture.SaveScreenshot(viewType, filePath);
                }
                else
                {
                    result = ScreenshotCapture.CaptureScreenshot(viewType, width, height);
                }
                
                if ((bool)result["success"])
                {
                    // Return proper MCP format with content array
                    if (result.ContainsKey("screenshots"))
                    {
                        var screenshots = result["screenshots"] as List<Dictionary<string, object>>;
                        if (screenshots != null && screenshots.Count > 0)
                        {
                            // Create MCP-compliant content array
                            var contentArray = new List<object>();

                            foreach (var screenshot in screenshots)
                            {
                                var base64Data = screenshot["data"] as string;
                                var mimeType = screenshot.ContainsKey("mimeType")
                                    ? screenshot["mimeType"] as string
                                    : "image/jpeg";

                                if (!string.IsNullOrEmpty(base64Data))
                                {
                                    contentArray.Add(new
                                    {
                                        type = "image",
                                        data = base64Data,
                                        mimeType = mimeType
                                    });
                                }
                            }

                            return JObject.FromObject(new
                            {
                                content = contentArray,
                                isError = false
                            });
                        }
                    }
                    else if (result.ContainsKey("files"))
                    {
                        return JObject.FromObject(new
                        {
                            content = new[]
                            {
                                new
                                {
                                    type = "text",
                                    text = result["message"] as string
                                }
                            },
                            isError = false
                        });
                    }
                }
                
                return JObject.FromObject(new
                {
                    content = new[]
                    {
                        new
                        {
                            type = "text",
                            text = result.ContainsKey("error") ? result["error"] as string : "Screenshot capture failed"
                        }
                    },
                    isError = true
                });
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"Screenshot tool error: {ex}");
                return JObject.FromObject(new
                {
                    content = new[]
                    {
                        new
                        {
                            type = "text",
                            text = $"Screenshot capture failed: {ex.Message}"
                        }
                    },
                    isError = true
                });
            }
        }
        
        private bool IsValidViewType(string viewType)
        {
            return viewType == "scene" || viewType == "game" || viewType == "both";
        }
    }
}